x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,488 @@
+		p.logger.Error(err)
+	}
+}
+
+func (p *Input) processorKeepAlive(svcSQS *sqs.Client, message sqs.Message, queueURL string, visibilityTimeout int64, errC chan error) {
+	for {
+		select {
+		case <-p.close:
+			return
+		case err := <-errC:
+			if err != nil {
+				p.logger.Warnf("Processing message failed: %v", err)
+				err := p.changeVisibilityTimeout(queueURL, visibilityTimeout, svcSQS, message.ReceiptHandle)
+				if err != nil {
+					p.logger.Error(errors.Wrap(err, "change message visibility failed"))
+				}
+				p.logger.Warnf("Message visibility timeout updated to %v", visibilityTimeout)
+			} else {
+				p.logger.Debug("ACK done, deleting message from SQS")
+				// only delete sqs message when errC is closed with no error
+				err := p.deleteMessage(queueURL, *message.ReceiptHandle, svcSQS)
+				if err != nil {
+					p.logger.Error(errors.Wrap(err, "deleteMessages failed"))
+				}
+			}
+			return
+		case <-time.After(time.Duration(visibilityTimeout/2) * time.Second):
+			// If half of the set visibilityTimeout passed and this is
+			// still ongoing, then change visibility timeout.
+			err := p.changeVisibilityTimeout(queueURL, visibilityTimeout, svcSQS, message.ReceiptHandle)
+			if err != nil {
+				p.logger.Error(errors.Wrap(err, "change message visibility failed"))
+			}
+			p.logger.Infof("Message visibility timeout updated to %v", visibilityTimeout)
+		}
+	}
+}
+
+func (p *Input) changeVisibilityTimeout(queueURL string, visibilityTimeout int64, svc *sqs.Client, receiptHandle *string) error {
+	req := svc.ChangeMessageVisibilityRequest(&sqs.ChangeMessageVisibilityInput{
+		QueueUrl:          &queueURL,
+		VisibilityTimeout: &visibilityTimeout,
+		ReceiptHandle:     receiptHandle,
+	})
+	_, err := req.Send(p.context)
+	return err
+}
+
+func getRegionFromQueueURL(queueURL string) (string, error) {
+	// get region from queueURL
+	// Example: https://sqs.us-east-1.amazonaws.com/627959692251/test-s3-logs
+	queueURLSplit := strings.Split(queueURL, ".")
+	if queueURLSplit[0] == "https://sqs" && queueURLSplit[2] == "amazonaws" {
+		return queueURLSplit[1], nil
+	}
+	return "", errors.New("queueURL is not in format: https://sqs.{REGION_ENDPOINT}.amazonaws.com/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
+}
+
+// handle message
+func handleSQSMessage(m sqs.Message) ([]s3Info, error) {
+	msg := sqsMessage{}
+	err := json.Unmarshal([]byte(*m.Body), &msg)
+	if err != nil {
+		return nil, errors.Wrap(err, "json unmarshal sqs message body failed")
+	}
+
+	var s3Infos []s3Info
+	for _, record := range msg.Records {
+		if record.EventSource == "aws:s3" && record.EventName == "ObjectCreated:Put" {
+			s3Infos = append(s3Infos, s3Info{
+				region: record.AwsRegion,
+				name:   record.S3.bucket.Name,
+				key:    record.S3.object.Key,
+				arn:    record.S3.bucket.Arn,
+			})
+		}
+	}
+	return s3Infos, nil
+}
+
+func (p *Input) handleS3Objects(svc s3iface.ClientAPI, s3Infos []s3Info, errC chan error) error {
+	s3Context := &s3Context{
+		errC: errC,
+	}
```

Comment:
Your initial count is 0. Depending on the 'situation' you either have a `-1`, or a too early 0.
Manual ref counting is a little like manual garbage collection. Someone must be the owner of s3Context at any point in time. Transfering ownership (the naive way) with ref counting means to be owner, increase refcount, transfer ownership, decrease ref count.

Reply:
Now after changing it to 1 on init, your function body is also owner of s3Context (which is good). But this means that your function body needs to relase ownership. Adding `defer s3Context.done()` should do the trick.